prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -0,0 +1,171 @@
+    :param pandas.DataFrame substations: data frame of substations.
+    :param int/float threshold: maximum mismatch distance (miles).
+    :return: (*pandas.DataFrame*) -- lines with matching substations.
+    """
+
+    def find_distance_to_closest_start_substation(line, substations):
+        # Coordinates are initially (lon, lat); we reverse to (lat, lon) for haversine
+        start_coordinates = line.loc["COORDINATES"][0][::-1]
+        start_name = line.loc["SUB_1"]
+        matching_start_substations = substations.loc[substations.NAME == start_name]
+        minimum_start_distance = matching_start_substations.apply(
+            lambda x: haversine(start_coordinates, (x.LATITUDE, x.LONGITUDE)), axis=1
+        ).min()
+        return minimum_start_distance
+
+    def find_distance_to_closest_end_substation(line, substations):
+        # Coordinates are initially (lon, lat); we reverse to (lat, lon) for haversine
+        end_coordinates = line.loc["COORDINATES"][-1][::-1]
+        end_name = line.loc["SUB_2"]
+        matching_end_substations = substations.loc[substations.NAME == end_name]
+        minimum_end_distance = matching_end_substations.apply(
+            lambda x: haversine(end_coordinates, (x.LATITUDE, x.LONGITUDE)), axis=1
+        ).min()
+        return minimum_end_distance
+
+    print("Evaluating endpoint location mismatches... (this may take several minutes)")
+    start_distance_mismatch = lines.apply(
+        lambda x: find_distance_to_closest_start_substation(x, substations), axis=1
+    )
+    end_distance_mismatch = lines.apply(
+        lambda x: find_distance_to_closest_end_substation(x, substations), axis=1
+    )
+    filtered = lines.loc[
+        (start_distance_mismatch > threshold) | (end_distance_mismatch > threshold)
+    ]
+    num_filtered = len(filtered)
+    num_lines = len(lines)
+    print(
+        f"dropping {num_filtered} lines with one or more substations with non-matching "
+        f"coordinates out of a starting total of {num_lines}"
+    )
+
+    remaining = lines.loc[
+        (start_distance_mismatch <= threshold) & (end_distance_mismatch <= threshold)
+    ]
+    return remaining.copy()
+
+
+def filter_lines_with_identical_substation_names(lines):
+    """Filter lines with SUB_1 or SUB_2 attributes equal to each other, and report the
+    number dropped.
+
+    :param pandas.DataFrame lines: data frame of lines.
+    :return: (*pandas.DataFrame*) -- lines with distinct substations.
+    """
+    num_lines = len(lines)
+    filtered = lines.query("SUB_1 == SUB_2")
+    num_filtered = len(filtered)
+    print(
+        f"dropping {num_filtered} lines with matching SUB_1 and SUB_2 out of a "
+        f"starting total of {num_lines}"
+    )
+    return lines.query("SUB_1 != SUB_2").copy()
+
+
+def build_transmission():
+    """Main user-facing entry point."""
+    # Load input data
+    hifld_substations = load.get_hifld_electric_substations(BLOB_PATH)  # noqa: F821
```

Comment:
Where is `BLOB_PATH` defined?

Reply:
It's not, and I didn't have the blob storage URL handy. I guess this is something that could be added to **const.py**.

Reply:
Yeah either that or just setting a placeholder value would be nice, so it's clear what the source is and how to override it. Whatever is easier.

Reply:
It will definitely be fixed before merging this branch. My goal was to get eyes on the rest of the code in the meantime.

Reply:
@danielolsen the location for the files are already in the annotation file:
https://github.com/Breakthrough-Energy/PreREISE/blob/hifld/ATTRIBUTION.md#grid

Reply:
Done.